etc/papd/print_cups.c

```diff
@@ -205,8 +193,199 @@ cups_printername_ok(char *name)         /* I - Name of printer */
+	 *    attributes-natural-language
+	 *    requested-attributes
+	 *    printer-uri
+	 */
+
+        request = ippNew();
+
+        ippSetOperation(request, IPP_GET_PRINTER_ATTRIBUTES);
+        ippSetRequestId(request, 1);
+
+	language = cupsLangDefault();
+
+	ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_CHARSET,
+		     "attributes-charset", NULL,
+		     cupsLangEncoding(language));
+
+	ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_LANGUAGE,
+		     "attributes-natural-language", NULL,
+		     language->language);
+
+	ippAddStrings(request, IPP_TAG_OPERATION, IPP_TAG_NAME,
+		      "requested-attributes",
+		      (sizeof(pattrs) / sizeof(pattrs[0])), NULL, pattrs);
+
+	ippAddString(request, IPP_TAG_OPERATION, IPP_TAG_URI,
+		     "printer-uri", NULL, uri);
+
+	/*
+	 * Do the request and get back a response...
+	 */
+
+        if ((response = cupsDoRequest(http, request, "/")) == NULL)
+        {
+                 LOG(log_error, logtype_papd,  "Unable to get printer attribs for %s - %s", name,
+                         ippErrorString(cupsLastError()));
+                httpClose(http);
+		cupsFreeDests(num_dests,dests);
+                return (0);
+        }
+
+        if (ippGetStatusCode(response) >= IPP_OK_CONFLICT)
+        {
+      		 LOG(log_error, logtype_papd,  "Unable to get printer attribs for %s - %s", name,
+                         ippErrorString(ippGetStatusCode(response)));
+                ippDelete(response);
+                httpClose(http);
+		cupsFreeDests(num_dests,dests);
+                return (0);
+        }
+
+	/*
+	* Range check input...
+	*/
+	
+	bufsize = sizeof(buffer);
+	if (buffer)
+		*buffer = '\0';
+
+	if (!buffer || bufsize < 1)
+	{
+		LOG(log_error, logtype_papd, "Check buffer failed!\n");
+		LOG(log_error, logtype_papd, strerror(EINVAL));
+                ippDelete(response);
+                httpClose(http);
+		cupsFreeDests(num_dests,dests);
+		return (0);
+	}
+
+	if (!response)
+	{
+		LOG(log_error, logtype_papd, "No IPP attributes.\n");
+                ippDelete(response);
+                httpClose(http);
+		cupsFreeDests(num_dests,dests);
+		return (0);
+	}
+
+	/*
+	* Open a temporary file for the PPD...
+	*/
+
+	if ((fp = cupsTempFile2(buffer, (int)bufsize)) == NULL)
+	{
+		LOG(log_error, logtype_papd, strerror(errno));
+                ippDelete(response);
+                httpClose(http);
+		cupsFreeDests(num_dests,dests);
+		return (0);
+	}
+
+	/*
+	 * Standard stuff for PPD file...
+	 */
+
+	cupsFilePuts(fp, "*PPD-Adobe: \"4.3\"\n");
+	cupsFilePuts(fp, "*FormatVersion: \"4.3\"\n");
+	cupsFilePrintf(fp, "*FileVersion: \"%d.%d\"\n", CUPS_VERSION_MAJOR, CUPS_VERSION_MINOR);
+	cupsFilePuts(fp, "*LanguageVersion: English\n");
+	cupsFilePuts(fp, "*LanguageEncoding: ISOLatin1\n");
+	cupsFilePuts(fp, "*PSVersion: \"(3010.000) 0\"\n");
+	cupsFilePuts(fp, "*LanguageLevel: \"3\"\n");
+	cupsFilePuts(fp, "*FileSystem: False\n");
+	cupsFilePuts(fp, "*PCFileName: \"ippeve.ppd\"\n");
```

Comment:
Would be helpful for posterity to have a comment explaining why this hardcoding is happening...

And, does this mean we cannot print text in encodings other than ISO Latin? I'll test tomorrow with KanjiTalk! :)

Reply:
This is a standard PPD file header. Most of it isn't actually used by papd. The Language options are relevant only to the PPD file and printer drivers that would read them (basically, "This file is in English, parse as such"). You can print all the KanjiTalk you want.